helix-term/src/commands/lsp.rs

```diff
@@ -424,81 +497,117 @@ impl ui::menu::Item for lsp::CodeActionOrCommand {
+    // this ensures, that a previously opened menu that doesn't have anything to do with this command will be replaced with a new menu
+    let code_actions_menu_open = Arc::new(Mutex::new(false));
 
-    let future = language_server.code_actions(
-        doc.identifier(),
-        range,
-        // Filter and convert overlapping diagnostics
-        lsp::CodeActionContext {
-            diagnostics: doc
-                .diagnostics()
-                .iter()
-                .filter(|&diag| {
-                    selection_range
-                        .overlaps(&helix_core::Range::new(diag.range.start, diag.range.end))
-                })
-                .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
-                .collect(),
-            only: None,
-        },
-    );
+    let mut requests = Vec::new();
 
-    cx.callback(
-        future,
-        move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
-            let actions = match response {
-                Some(a) => a,
-                None => return,
-            };
-            if actions.is_empty() {
-                editor.set_status("No code actions available");
-                return;
-            }
+    for language_server in doc.language_servers_with_feature(LanguageServerFeature::CodeAction) {
+        let offset_encoding = language_server.offset_encoding();
+        let range = range_to_lsp_range(doc.text(), selection_range, offset_encoding);
+
+        requests.push((
+            language_server.code_actions(
+                doc.identifier(),
+                range,
+                // Filter and convert overlapping diagnostics
+                lsp::CodeActionContext {
+                    diagnostics: doc
+                        .diagnostics()
+                        .iter()
+                        .filter(|&diag| {
+                            selection_range
+                                .overlaps(&helix_core::Range::new(diag.range.start, diag.range.end))
+                        })
+                        .map(|diag| diagnostic_to_lsp_diagnostic(doc.text(), diag, offset_encoding))
+                        .collect(),
+                    only: None,
+                },
+            ),
+            offset_encoding,
+            language_server.id(),
+        ));
+    }
+
+    for (future, offset_encoding, lsp_id) in requests {
+        let code_actions_menu_open = code_actions_menu_open.clone();
+
+        cx.callback(
+            future,
+            move |editor, compositor, response: Option<lsp::CodeActionResponse>| {
+                let actions = match response {
+                    Some(a) => a
+                        .into_iter()
+                        .map(|a| (a, offset_encoding))
+                        .collect::<Vec<_>>(),
+                    None => return,
+                };
+
+                let mut code_actions_menu_open = code_actions_menu_open.lock().unwrap();
```

Comment:
Should probably just show an error if the lock is poisoned.

Reply:
Hmm I don't know (makes code certainly less readable).
I think if the lock is poisoned, something is really wrong already. I guess the editor would crash somewhere else or has already crashed in one way or the other before.

There are whole discussions (e.g. here: https://old.reddit.com/r/rust/comments/wvynot/does_rust_have_any_design_mistakes/ilimprg/) about why returning a `Result` in `lock` by default may have been a wrong design decision)

Reply:
On a second thought, I handled this now since it doesn't make the code *that* much less readable. But I still think this is probably unnecessary (and unwrapping is fine here).

Reply:
Yeah, you're probably right. I have no strong feelings about this.